### PR TITLE
Xcode 10: Support Swift 4 & Swift 4.2

### PR DESCRIPTION
**Pull request by freak4pc**

### CODE REVIEW

**Comment by freak4pc** [comment]
Not sure why CI isn't running. Could be that forks aren't turned on in settings 🤔 
cc/ @ashfurrow

**Comment by ashfurrow** [comment]
Oh haha, well it's probably since this repo is still on Circle CI 1.0 😅 I'll work on upgrading it and then we can merge this 👍

**Comment by freak4pc** [comment]
Need any help with v2 migration or you'll take care of it? :]

**Comment by ashfurrow** [comment]
Nah, thanks for offering though! I'd like to go through the process again – it's been a few months 😅

**Comment by freak4pc** [comment]
You got it !
Here's some reference if you want it ;-) 
https://github.com/RxSwiftCommunity/RxSwiftExt/blob/master/.circleci/config.yml

**Comment by ashfurrow** [comment]
Cool, thanks! Can you pull from master and see if this passes on CI now?

**Comment by freak4pc** [comment]
Yes sir! :]

**Comment by freak4pc** [comment]
Rebased over master, let's see! 🤞

**Comment by freak4pc** [comment]
Ugh that's kinda strange. Works fine locally with Xcode 10 GM and not sure what Xcode 10 Circle is running. I'll try redoing the snapshots later to see what's going on.

**Comment by ashfurrow** [comment]
Hmm! Not sure :\ Could be a good use case for [second_curtain](https://github.com/ashfurrow/second_curtain) 😄

File: Rakefile [hunk 1]
```diff
@@ -1,8 +1,9 @@
 require 'tmpdir'
 
-desc 'Run unit tests on iOS 11.2'
+desc 'Run unit tests on Swift 4.0 & Swift 4.2'
 task :test do
-  sh "set -o pipefail && xcodebuild -workspace 'Bootstrap/Bootstrap.xcworkspace' -sdk 'iphonesimulator' -scheme 'Bootstrap' -destination 'name=iPhone 6,OS=11.2' build test | xcpretty --color --simple"
+  sh "set -o pipefail && xcodebuild -workspace 'Bootstrap/Bootstrap.xcworkspace' -sdk 'iphonesimulator' -scheme 'Bootstrap' -destination 'name=iPhone 8' SWIFT_VERSION=4.0 build test | xcpretty --color --simple"
+  sh "set -o pipefail && xcodebuild -workspace 'Bootstrap/Bootstrap.xcworkspace' -sdk 'iphonesimulator' -scheme 'Bootstrap' -destination 'name=iPhone 8' SWIFT_VERSION=4.2 build test | xcpretty --color --simple"
```
Comment (ashfurrow) [inline Rakefile]: We'll need to update these to look like this:

```sh
set -o pipefail && xcodebuild ... | second_curtain | xcpretty ...
```

Notice how it pipes the output through `second_curtain` _before_ it goes into `xcpretty`. This has to be there to work.

Comment (ashfurrow) [reply Rakefile]: Oh shoot, I meant to leave this on #165 oops.

Comment (freak4pc) [reply Rakefile]: _cries hard_

**Comment by Vkt0r** [comment]
@freak4pc Can you pull from master to run in the CI now with `second_curtain` integrated and we can see what's the problem ?

**Comment by freak4pc** [comment]
Sure. I'm 99% sure it's unrelated to a visual change (just OS differences), but I'll try :)

**Comment by ashfurrow** [comment]
The new changes have the output of `xcodebuild` saved to a file and read later, which won't work with running them twice like this.

I say, try it and see, but I'm fine with only testing just one specific iOS version in CI.

**Comment by freak4pc** [comment]
It's not about a specific iOS version, It's about making sure it builds on both Swift 4 and Swift 4.2. Let's see how it works.

**Comment by freak4pc** [comment]
Uhm... it... passes? O.o 

Kay! 🤷‍♂️

**Comment by Vkt0r** [comment]
@freak4pc I'm missing something here 🤯😂?

**Comment by Vkt0r** [comment]
Shame, I was hoping to put my hands in that images 😂🤣.

**Comment by ashfurrow** [comment]
Sure, we could do that! We can go even further, depending on who is interested in doing this. We could use a workflow to define two parallel jobs with different Swift versions, maybe each job would just specify a different `environment`? I'm not saying it's necessary, or even a good idea, but if anyone is interested in playing with CI, this repo could serve as a fun place to test ideas.

**Comment by freak4pc** [comment]
CircleCI free path doesn't do parallelism anyways so we won't be benefiting from workflows (would actually take longer because of bootstrapping).

**Comment by ashfurrow** [comment]
Argh boooo 😭

**Comment by Vkt0r** [comment]
Yes you're right they don't support it for macOS project, they do for Linux I think.

**Comment by Vkt0r** [comment]
@ashfurrow @freak4pc I think we can create for now separate jobs for each version of swift and add handle it the xcodebuild log with `second_curtain` for now for each one separately too. What do you think guys ? @freak4pc You take the challenge from here as is your PR ? 😅🙏

**Comment by ashfurrow** [comment]
It does feel cleaner, and it would probably scale nicely.

**Comment by freak4pc** [comment]
Yup, as suggested earlier I believe separate jobs would work better here.
I'm flying out to Germany to talk in SwiftConf so I'll be able to look into this Fri-Sat probably. 

If you want to take over, feel free ! :]

**Comment by ashfurrow** [comment]
Have fun!

**Comment by Vkt0r** [comment]
Have fun in SwiftConf! Don't worry I'll take over it during the rest of week as soon as I can .

**Comment by freak4pc** [comment]
@Vkt0r Are you still interested in taking ownership here or should I find some time for this?

**Comment by Vkt0r** [comment]
@freak4pc Sorry I've been really busy since we talked about this. I set up something but I haven't finished yet. Tomorrow or maybe tonight I'll finish it and push it. By the way, I'm not sure if I have access to push in your branch by default I think 🤔.

**Comment by freak4pc** [comment]
Uhm I'm not 100% sure, I believe it should let you as you're a Collaborator on the repo. Anyways if you'd like me to take care of it just say the word. Cheers!

**Comment by Vkt0r** [comment]
@freak4pc @ashfurrow I created a separate PR based in this one with the modifications to CircleCI (#167) to test it isolated. Just a few thoughts about it:

- I thought about to remove the rake test and just run it as steps from CircleCI, but I think have the Rakefile independent is really useful to test locally (at least for me is really useful every time I create a PR 😅).
- I set the `test` task in the `Rakefile` to use the ENV variable injected in CircleCI, please let me know your thoughts about it.

With another PR we can test it and change it as we like and then when is ready @freak4pc you can integrate the two files in your PR 😀. Please take a look iat#167 and let me know anything you think we can improve.

**Comment by Vkt0r** [comment]
@ashfurrow @freak4pc Did you have time to see what I mentioned above 😁 ?. As soon as we agree with that we can integrate it here and update the project 😅. Cheers

**Comment by freak4pc** [comment]
I’m personally quite fine with getting rid of the Rakefile ... I don’t see so much value in such a small project given we’ll need to repeat some of the expressions in the CI. We can go either way though, leaving this to ash as it’s his project. 

It’s holiday season here so a bit harder for me to take a look but I’ll try my best over the next few days / weekend. 

Thanks for all your work ! 
Shai.

**Comment by ashfurrow** [comment]
Yup, agreed. Thanks for all your work – feel free to get rid of the Rakefile or do whatever is necessary 👍

**Comment by Vkt0r** [comment]
@freak4pc @ashfurrow Thanks for your quick reply guys 👏😁. Sure I'll get rid of the **Rakefile** and continue testing in the another PR. Enjoy your holiday season Shai 🎉.

**Comment by Vkt0r** [comment]
Guys I think we are ready to go with #167, please take a look and let me know if you find anything we can improve. I removed the Rakefile as we agreed and I added the `swiftlint` step too, in this moment is failing, but I prefer to fix all the warning and errors in another PR specifically for that 😅. As soon as we agree we can bring the changes in the `config.yml` and `Rakefile` to this PR and merge it 🎉.

**Comment by Vkt0r** [comment]
Thanks for the update Shai! Please take your time, no rush 😀. Cheers !

**Comment by Vkt0r** [comment]
By the way guys, this sounds awesome by the guys of CircleCI ([Reusing config](https://circleci.com/docs/2.0/reusing-config/?mkt_tok=eyJpIjoiTkdVME5qQXlObU13TlRJeCIsInQiOiJQV1VkcUs2NjA2dnZwMHJibnM4ODVEMWMwTFpnbmxZN1c4K0lPaGw2TTVKRDJwVVwvckw4YUlVUEZYdllSZVR3OXhTUWVkT3JmNUprY2pKeVhIQjEzTHZjdW9TVktPbytoVlY0a3FnMXNjSmZYbzFrbXFPSVNYQ3J2OURRXC9VTWQxIn0%3D#section=configuration)), so maybe I'll make the change to refactor some repetition in the commands 😅 before we merge it.

**Comment by Vkt0r** [comment]
@freak4pc I committed the changes of the previous PR in yours to consolidate it as the Swift4.2 update. We have the Swiftlint step failing regarding some warnings. I think we can create another PR to fix it later

**Comment by Vkt0r** [comment]
@ashfurrow I think this is ready to be reviewed when you have time 😅. I'll come back with another PR to fix all the Swiftlint issues later after this has been merged if works for you guys.

**Comment by freak4pc** [comment]
Gonna be able to review this tomorrow finally ! :)

**Comment by Guerrix** [comment]
Hello there, first of all thanks for all your hard work. Do you have an ETA when this PR will be merged?. Thanks

**Comment by freak4pc** [comment]
Hey @Vkt0r , Thank you so much for the work. Just had the change to look over this.

Overall this looks good, there are just two issues I can see:

1. SwiftLint fails remotely but works locally. This either means a difference between versions, or the CI is fetching some local cache. 

2. Running 4 separate jobs for each build isn't the best idea on a free plan. I believe that's not the plan we discussed initially, and if yes I must've misunderstood. The point was having a single job that runs all 4 steps sequentially so you don't have to wait for 4 separate jobs, where each has to do its own checkout and build. 

Let me know your thoughts. 

Thanks!
Shai.

**Comment by Vkt0r** [comment]
Hi Shai, you're very welcome, thanks to you too for this hard work! I went through the mentioned issues:

>SwiftLint fails remotely but works locally...

Uhmm I think this is very weird, it fails for me locally running `swiftlint lint --no-cache` in the root directory of the project.  I can confirm my version of SwiftLint is `0.27.0`. Nevertheless, I think we can come later with another PR to fix the warnings and add Swiftlint as a Build Phase Script in the project to prevent it in Xcode 🤔.

>Running 4 separate jobs for each build isn't the best idea on a free plan. I believe that's not the plan we discussed initially, and if yes I must've misunderstood. The point was having a single job that runs all 4 steps sequentially so you don't have to wait for 4 separate jobs, where each has to do its own checkout and build.

Agree with you that running 4 separate jobs in the free plan can be a little painful, even when in this case there are small jobs except the `tests`. I don't have any problem with running all the jobs as one single job that in fact could be faster as we should need to check out the project only once (I think we could do this using the `workspaces` too).

IMO the advantage I can see of separate the jobs is having a more meaningful feedback of which jobs are failing without to go to the CI and see all the logs, a better way to isolate it. But again is in IMHO 😅.

As far as we agree what we want to do so we can update the PR 😀. Let me know guys what do you think about it.
CC @freak4pc, @ashfurrow

**Comment by Vkt0r** [comment]
@Guerrix We would try to merge it this week 🤞😄.

**Comment by freak4pc** [comment]
@Vkt0r IRT to SwiftLint, my bad, it should be fine now. I prefer fixing everything in this PR as I don't feel comfortable merging something with a failing build step (other option would be removing that build step and then adding it separately in another PR)

IRT to separate steps - I prefer one build process in this case, but I'm OK with whatever @ashfurrow feels comfortable with. 

Overall this works, and seems to work well - I just feel this would cause the build process to be longer because of the bootup, "waiting" and loading of each separate build process. 

Again, major kudos for the hard work!

**Comment by Vkt0r** [comment]
@freak4pc Hahaha I was ready to push exactly the same changes 😂. Kudos for you too ! Sure I'm OK too with whatever Ash feels comfortable with 😀. Thanks again 👍

**Comment by ashfurrow** [comment]
One build step sounds like the best idea, let’s go with that 👍 Thanks everybody!

**Comment by paulz** [comment]
👍

**Comment by freak4pc** [comment]
Hey @paulz  - there's already an open PR. Any reason you opened another one? 

Also, @Vkt0r - will you be able to take care of the single step or should I ? I would really like to see this PR through by the weekend if somehow possible as its been open for over 3 weeks now. 

Thanks!

**Comment by Vkt0r** [comment]
@freak4pc Sorry I was traveling all the weekend definitely I will do the change to a single step ASAP today or maybe tomorrow.

**Comment by freak4pc** [comment]
No need to apologize at all ! Just wanted to offer my help if you need it.
Thanks !

**Comment by Vkt0r** [comment]
@freak4pc I modified it to run it as single job as we agreed with Ash. Take a look and let me know

**Comment by freak4pc** [comment]
Perfect! Only one nit-pick, this line in your YAML doesn't work as you expdect:
`tee output/xcodebuild_swift$SWIFT_VERSION.log` 

Look in your artifcats, the swift_version isn't part of the output: 

image

**Comment by freak4pc** [comment]
@Vkt0r Actually I don't think that ENV var exists - I would just hardcode `xcodebuild_swift4.log` and `xcodebuild_swift4_2.log`.

**Comment by Vkt0r** [comment]
@freak4pc Upps that wasn't cleaned 😅. Clean it now, good catch 👍.

**Comment by freak4pc** [comment]
LGTM @ashfurrow

**Comment by ashfurrow** [comment]
Hmm, this broke CI on master, the `second_curtain` invocation fails. I've opened #170 to track.

**Pull request merged**